python/TestHarness/tests/test_QueuePBS.py

```diff
@@ -63,100 +51,364 @@ def createHarnessAndTesters(self, test_file):
+            # remove previous session data and the session file
+            self._clean()
+            os.remove(self.pbs_session_file)
+        except:
+            pass
+
+    def tearDown(self):
         """
-        (harness, testers) = self.createHarnessAndTesters(self.test_file)
-        with mock.patch.object(util, 'runCommand') as mock_run:
-            # A proper return value for qsub command
-            mock_run.return_value = '1.nowhere.org'
+        tearDown occurs after every test.
 
-            # Instruct the TestHarness to schedule tests
-            harness.scheduler.schedule(testers)
-            harness.scheduler.waitFinish()
+        Clean up old session file and sub-directories.
+        """
+        try:
+            # remove previous session data and the session file
+            self._clean()
+            os.remove(self.pbs_session_file)
+        except:
+            pass
 
-        return mock_result.call_args
+    def _clean(self):
+        """ clean the session sub-directories only (leave session file intact) """
+        if os.path.exists(self.session_dir):
+            try:
+                shutil.rmtree(self.session_dir)
+            except:
+                raise Exception('Failed to clean up pre-existing session directories')
 
     @mock.patch.object(util, 'formatResult')
-    def _mockPBSStatus(self, qstat_status, mock_result):
+    def _mockSinglePBSLaunch(self, harness, job, fake_result, mock_result):
+        """
+        Launch 'job' and mock the output of runCommand to 'fake_result', then
+        return a list of the last thing(s) the TestHarness tried to print.
+
+        Note: Because we are wanting to control each launched job, we must
+        mock getNextJobGroup. But by doing so we need to become the Scheduler's
+        job manager per say. Hence why, we are checking and modifying the
+        tester's status.
+        """
+
+        # Limit printing of MagicMock
+        mock_result.return_value = ''
+
+        with mock.patch.object(harness.scheduler, 'getNextJobGroup') as mock_next_jobs:
+            with mock.patch.object(util, 'runCommand') as mock_run:
+                # Setup Mock returns
+                mock_run.return_value = fake_result
+                mock_next_jobs.return_value = []
+                tester = job.getTester()
+
+                if tester.isInitialized():
+                    tester.setStatus('PENDING', tester.bucket_pending)
+
+                # Skipped jobs are sent to a different method and have already
+                # been processed during the iteration that created this
+                # condition in the first place (jobA fails and thus skips and
+                # never runs jobB, but _occurres_ during jobA's iteration).
+                elif tester.isSkipped():
+                    return
+
+                # Launch the job and wait (non-blocking)
+                harness.scheduler.queueJobs(run_jobs=[job])
+                while tester.isPending():
```

Comment:
This is okay, because the Scheduler is using thread timers to handle timeouts. If we happen to hang here, it will generate a TIMEOUT error.